analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs

```diff
@@ -27,122 +28,110 @@
+                    : c.SemanticModel.GetDeclaredSymbol(variableSyntax);
+                if (variableSymbol != null)
+                {
+                    symbolsWhereTypeIsCreated.Add(variableSymbol.ToSymbolWithSyntax(c.Node));
+                }
+            };
+
+        private static Action<SyntaxNodeAnalysisContext> ProcessSimpleAssignments(ConcurrentBag<ISymbol> symbolsWhereLocaleIsSet) =>
+            c =>
             {
-                return assignment.Left;
-            }
+                var assignmentExpression = (AssignmentExpressionSyntax)c.Node;
+                var propertySymbol = GetPropertySymbol(assignmentExpression, c.SemanticModel);
 
-            if (parent is EqualsValueClauseSyntax equalsClause &&
-                equalsClause.Parent.Parent is VariableDeclarationSyntax variableDeclaration)
+                if (propertySymbol != null
+                    && propertySymbol.ContainingType.IsAny(CheckedTypes)
+                    && propertySymbol.Name == "Locale")
+                {
+                    var variableSymbol = GetAccessedVariable(assignmentExpression, c.SemanticModel);
+                    if (variableSymbol != null)
+                    {
+                        symbolsWhereLocaleIsSet.Add(variableSymbol);
+                    }
+                }
+            };
+
+        private static Action<CompilationAnalysisContext> ProcessCollectedSymbols(ConcurrentBag<SyntaxNodeWithSymbol<SyntaxNode, ISymbol>> symbolsWhereTypeIsCreated,
+                                                                                  ConcurrentBag<ISymbol> symbolsWhereLocaleIsSet) =>
+            c =>
             {
-                return variableDeclaration.Variables.Last();
-            }
+                foreach (var invalidCreation in symbolsWhereTypeIsCreated.Where(x => !symbolsWhereLocaleIsSet.Contains(x.Symbol)))
+                {
+                    var typeName = invalidCreation.Symbol.GetSymbolType()?.Name;
+                    if (typeName == null)
+                    {
+                        continue;
+                    }
 
-            return null;
-        }
+                    c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, invalidCreation.Syntax.GetLocation(), typeName));
+                }
+            };
 
-        private static IPropertySymbol GetPropertySymbol(AssignmentExpressionSyntax assignment, SemanticModel model)
-        {
-            return model.GetSymbolInfo(assignment.Left).Symbol as IPropertySymbol;
-        }
+        private static ISymbol GetSymbolFromConstructorInvocation(SyntaxNode constructorCall, SemanticModel semanticModel) =>
+            constructorCall is ObjectCreationExpressionSyntax objectCreation
+                ? semanticModel.GetSymbolInfo(objectCreation.Type).Symbol
+                : semanticModel.GetSymbolInfo(((ImplicitObjectCreationExpressionSyntaxWrapper)constructorCall).SyntaxNode).Symbol?.ContainingType;
```

Comment:
`((ImplicitObjectCreationExpressionSyntaxWrapper)constructorCall).SyntaxNode` is actually `constructorCall` itself. So we don't need the indirection :) 
```suggestion
                : semanticModel.GetSymbolInfo(constructorCall).Symbol?.ContainingType;;
```